ember-client/shared/src/main/scala/org/http4s/ember/client/internal/WebSocketHelpers.scala

```diff
@@ -37,6 +54,78 @@ private[client] object WebSocketHelpers {
   private[internal] val webSocketProtocol = Protocol(ci"websocket", None)
   private[internal] val connectionUpgrade = Connection(NonEmptyList.of(upgradeCi))
   private[internal] val upgradeWebSocket = Upgrade(webSocketProtocol)
+  private[internal] val exampleSecWebSocketKey = "dGhlIHNhbXBsZSBub25jZQ=="
+
+  // private[this] val clientTranscoder = new FrameTranscoder(true)
+
+  private[internal] def createWebSocketKey[F[_]]: Key[Socket[F]] =
+    Key.newKey[SyncIO, Socket[F]].unsafeRunSync()
+
+  def getSocket[F[_]](client: Client[F], request: Request[F])(implicit
+      F: MonadCancel[F, Throwable]
+  ): Resource[F, Resource[F, Option[Socket[F]]]] = {
+    val webSocketKey = createWebSocketKey[F]
+    client
+      .run(request)
+      .map { res =>
+        Resource
+          .eval(validateServerHandshake(res, exampleSecWebSocketKey))
+          .map(isValid => isValid.toOption *> res.attributes.lookup(webSocketKey))
+      }
+  }
```

Comment:
The type could be simplified a bit here
```suggestion
  def getSocket[F[_]](client: Client[F], request: Request[F])(implicit
      F: MonadCancel[F, Throwable]
  ): Resource[F, Option[Socket[F]]] = {
    val webSocketKey = createWebSocketKey[F]
    client
      .run(request)
      .flatMap { res =>
        Resource
          .eval(validateServerHandshake(res, exampleSecWebSocketKey))
          .map(isValid => isValid.toOption *> res.attributes.lookup(webSocketKey))
      }
  }
```